### PR TITLE
Explain gas implications of constant and immutable variables.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: docs/contracts/constant-state-variables.rst [hunk 1]
```diff
@@ -12,6 +12,14 @@ for ``immutable``, it can still be assigned at construction time.
 The compiler does not reserve a storage slot for these variables, and every occurrence is
 replaced by the respective value.
 
+Compared to regular state variables, the gas costs of constant and immutable variables
+are much lower. For a constant variables, the expression assigned to it is copied to
```
Comment (bshastry) [inline docs/contracts/constant-state-variables.rst]:
```suggestion
are much lower. For constant variables, the expression assigned to it is copied to
```

File: docs/contracts/constant-state-variables.rst [hunk 2]
```diff
@@ -12,6 +12,14 @@ for ``immutable``, it can still be assigned at construction time.
...
+all the places where it is accessed and also re-evaluated each time. This allows for local
+optimizations. Immutable variables are evaluated once at construction time and their value
+is copied to all the places where they are accessed. For this, 32 bytes are reserved
+for the value, even if it is smaller. Because of that, constant values could be cheaper
+than immutable value.
```
Comment (bshastry) [inline docs/contracts/constant-state-variables.rst]:
```suggestion
than immutable values.
```

File: docs/contracts/constant-state-variables.rst [hunk 3]
```diff
@@ -12,6 +12,14 @@ for ``immutable``, it can still be assigned at construction time.
 The compiler does not reserve a storage slot for these variables, and every occurrence is
 replaced by the respective value.
 
+Compared to regular state variables, the gas costs of constant and immutable variables
+are much lower. For a constant variables, the expression assigned to it is copied to
+all the places where it is accessed and also re-evaluated each time. This allows for local
+optimizations. Immutable variables are evaluated once at construction time and their value
+is copied to all the places where they are accessed. For this, 32 bytes are reserved
```
Comment (bshastry) [inline docs/contracts/constant-state-variables.rst]:
```suggestion
is copied to all the places where they are accessed. 32 bytes are reserved
```

File: docs/contracts/constant-state-variables.rst [hunk 4]
```diff
@@ -12,6 +12,14 @@ for ``immutable``, it can still be assigned at construction time.
...
+for the value, even if it is smaller. Because of that, constant values could be cheaper
```
Comment (bshastry) [inline docs/contracts/constant-state-variables.rst]:
```suggestion
to hold the value of an immutable variable, even if its assigned value is smaller. Due to this, constant values could be cheaper
```

File: docs/contracts/constant-state-variables.rst (hunk 1 above)
Comment (leonardoalt) [inline docs/contracts/constant-state-variables.rst]:
```suggestion
are much lower. For a constant variable, the expression assigned to it is copied to
```

**Review by bshastry — COMMENTED** [review]
Minor comments

File: docs/contracts/constant-state-variables.rst (hunk 2 above)
Comment (leonardoalt) [inline docs/contracts/constant-state-variables.rst]:
```suggestion
than immutable values.
```

**Review by leonardoalt — CHANGES_REQUESTED** [review]

File: docs/contracts/constant-state-variables.rst (hunk 3 above)
Comment (chriseth) [reply docs/contracts/constant-state-variables.rst]: I try to avoid starting a sentence with a number, so I did a different change.

**Review by bshastry — APPROVED** [review]
LGTM

**Review by Marenz — APPROVED** [review]

**Pull request merged**